WordPress/gutenberg
# Avoid persisting preference every time the sidebar tab is changed

## What?
Attempts to solve the feedback here - https://github.com/WordPress/gutenberg/pull/39795#issuecomment-1111883457

> Anyway, it does feel like switching between sidebars shouldn't be a preference at all. Whether the sidebar is closed or open should be, so there's a small refactoring to be done here to improve things.

## Why?
For a long time, we've been saving which type of sidebar is active (document, block, global styles, plugin sidebar) as a preference.

This was ok when the preferences were only persisted in local storage. Since #39795, persisting preferences has involved an HTTP request, and this is potentially expensive for such a regular action as changing which sidebar is visible.

## How?
This introduces less granular preferences state `isComplementaryAreaVisible`, which is only set when the sidebar is closed or opened. The type/name of the sidebar that's open is now stored only in-memory using a reducer in the interface store.

## Drawbacks

- This does mean if the user had Global Styles or the Navigation sidebar open in the site editor, when they reload the page the 'document' sidebar will instead be open.

- There's a situation where a sidebar/complementary area can be open, but we don't know exactly which type of sidebar to display (because the 'area' is no longer persisted). In this case I've introduced the `setDefaultComplementaryArea` action, which is used when initializing the site editor (and used in the site editor, which doesn't show a sidebar by default).

## Testing Instructions

1. Open an editor
2. Close the sidebar on the right
3. Reload
4. The sidebar on the right should still be closed
5. Expand the sidebar on the right
6. Reload
7. The sidebar on the right should still be expanded.
